net: sockets: dtls: reset mbedtls session on timed out handshake

emillindq opened the pull request:
This PR fixes #35711

According to MbedTLS API documentation, its session must be reset if mbedtls_ssl_handshake returns timeout error. This commit resets the session for said return value, and that allows us to call send() multiple times even if handshake times out for previous calls.

Please note that `k_sem_init` resets the semaphores internal list of polls waiting for it, and therefore a poll waiting on the `tls_established` sem will block forever if first handshake fails. This was changed to `k_sem_reset` in this PR.

This PR is a draft because I wanted to ask @rlubos why the `context->dtls_peer_addr` is erased in `tls_mbedtls_reset`? This breaks stuff since we then forget which our peer is, which is required if calling first `connect` and then `send`, [dtls hs timeout], `send` (which the echo_server/client does). What I mean is, is this strictly required in some cases? Thanks!

emillindq committed:
net: sockets: dtls: reset mbedtls session on timed out handshake

According to MbedTLS API documentation, its session must be
reset if mbedtls_ssl_handshake returns timeout error. This
commit resets the session for said return value, and that
allows us to call send() multiple times even if handshake
times out for previous calls.

Fixes #35711

Signed-off-by: Emil Lindqvist <emil@lindq.gr>

rlubos commented:
> This PR is a draft because I wanted to ask @rlubos why the context->dtls_peer_addr is erased in tls_mbedtls_reset? This breaks stuff since we then forget which our peer is, which is required if calling first connect and then send, [dtls hs timeout], send (which the echo_server/client does). What I mean is, is this strictly required in some cases? Thanks!

This is needed for the DTLS server, where you don't `connect()` explicitly. Once some client establishes connection with a DTLS server, `dtls_peer_addr` is set to its address, so that we don't process packets from different peers (we only support a single DTLS session per socket). After the client ends the session, we reset the context **and** the address so that a new one can connect w/o a need to reopen a socket.

Since it shouldn't be needed for a client to clear the peer address, I guess it should be enough to verify the DTLS role in the `tls_mbedtls_reset()` function.

emillindq commented:
> Since it shouldn't be needed for a client to clear the peer address, I guess it should be enough to verify the DTLS role in the `tls_mbedtls_reset()` function.

Thanks for this feedback; I have changed and added a comment about this.

I do feel quite confident in this PR; I have tested it with the following setup
1. Two stm32h747i_disco_m7 devkits: client and server
2. Hold button on server on boot to make it enter the network (aka. reply to ARP) but not reply to UDP/TCP.
3. Client and server is started with server put into mode above. TLS will fail and client will fail DTLS handshake over and over.
4. After a while server is rebooted with normal mode
5. We watch DTLS handshake succeed and echoing begins on UDP. TLS is not recovered since the connect() TCP call failed.

Tested on both IPV4 & IPV6. Please see https://github.com/emillindq/zephyr/tree/tls_handshake_reset_test for source.

rlubos commented:
> 5\. We watch DTLS handshake succeed and echoing begins on UDP. TLS is not recovered since the connect() TCP call failed.

And I think that's what should be expected, right? The man pages for `connect()` are clear, that if the `connect()` call fails, one should close/reopen the socket before another attempt:

```
       If connect() fails, the state of the socket is unspecified.
       Conforming applications should close the file descriptor and
       create a new socket before attempting to reconnect.
```

rlubos reviewed (commented):
Looks good, just one nit.

emillindq commented on subsys/net/lib/sockets/sockets_tls.c:
```diff
@@ -809,12 +809,19 @@ static int tls_mbedtls_reset(struct tls_context *context)
+	 */
+	if (context->options.role == MBEDTLS_SSL_IS_SERVER) {
+		(void)memset(&context->dtls_peer_addr, 0,
+		sizeof(context->dtls_peer_addr));
```
shoot, that was a function!

rlubos reviewed (approved):
LGMT, thanks!
